addons/website_sale_loyalty/models/loyalty.py

```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, tools
+from odoo.tools.safe_eval import safe_eval
+
+
+class LoyaltyProgram(models.Model):
+    _name = 'loyalty.program'
+    _inherit = ['loyalty.program']
+
+    def get_available_rewards(self, order):
+        """ Returns the list of rewards that the cart can get """
+        client = order.partner_id
+        rewards = []
+        if client:
+            spendable_points = tools.float_round(client.loyalty_points - order.spent_loyalty_points, 0, rounding_method='HALF-UP')
+            for reward in self.reward_ids:
+                if reward.minimum_points > spendable_points:
+                    continue
+                elif reward.reward_type == 'discount' and reward.point_cost > spendable_points:
+                    continue
+                elif reward.reward_type == 'gift' and reward.point_cost > spendable_points:
+                    continue
+                elif reward.reward_type == 'discount' and reward.discount_apply_on == 'specific_products':
+                    if not (order.order_line.mapped('product_id') & reward.discount_specific_product_ids):
+                        continue
+                elif reward.reward_type == 'discount' and reward.discount_type == 'fixed_amount' and order.amount_total < reward.minimum_amount:
+                    continue
+                elif reward in order.order_line.mapped('loyalty_reward_id'):
+                    continue
+                rewards.append(reward)
+        return rewards
+
+    def contains_discount_reward(self, product_id):
+        """Returns true if this reward contains the specified product_id as a discount reward"""
+        for reward in self.reward_ids:
+            if reward.reward_type == 'discount' and reward.discount_product_id.sudo().product_tmpl_id.id == product_id.id:
+                return True
+        return False
```

Comment:
As the first condition is cheaper than the second one, I suggest to test it on all records first.
For the second condition, you may take advantage of the automatic "mapped" on relational fields.
```suggestion
        return any(
            reward.reward_type == 'discount' for reward in self.reward_ids
        ) or (
            product_id in self.reward_ids.discount_product_id.sudo().product_tmpl_id
        )
```

Reply:
Unfortunately both conditions need to apply to the same reward: if reward1 matches condition 1 and reward2 matches condition 2, it should return false.
However I can do this instead:
```py
        return self.reward_ids.filtered(
            lambda reward: reward.reward_type == 'discount' and product_id in reward.discount_product_id.sudo().product_tmpl_id
        )
```